design/kiota-e2e.md

```diff
@@ -0,0 +1,147 @@
+# End-to-end usage of kiota-generated SDKs in Javascript
+
+While working with the Kiota + Javascript teams, it became clear that we needed more context and provide a better vision on how we should be using a kiota-generated SDK. This design document aims at providing clarity and a better understanding on how developers will be using our SDKs and the underlying kiota packages.
+
+## Constraints
+
+Before we jump into the end-to-end walk-through, it's important to set some constraints.
+
+| Type      | Description                                                                                                                                   |
+| --------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
+| Platforms | Node : Current and Previous LTS (14 / 16)<br /> Web : Edge, Chrome, Firefox and Safari (Latest released version + immediate previous version) |
+| Modules   | Node : CommonJS<br /> Web : ES Module (ES6)                                                                                                   |
```

Comment:
just curious to what data point lead to commonjs for node?

Reply:
I made the decision based on current "state" of NodeJS... We're still seeing way more CommonJS modules in the wild today and we need to be where the developers are. Right now, ES Modules are still marked as "experimental" in Node 14 LTS but not in NodeJS 16 LTS... For that reason, I don't want to build and having to refactor based on the changes they could make.

https://nodejs.medium.com/node-js-version-14-available-now-8170d384567e

That being said, the conversation is different with NodeJS 16. This is fully supported and is the path forward.

Now my question to you @nikithauc and @baywet : As we're planning to build both (CJS and ESM), could we focus on ESM for now and add CJS in the future if the data tells us that we should? What would be the downside to build on LTS 16 and see how this works?

Reply:
I believe that we should be able add CJS support later should we need it. We're talking about a different TS configuration, some orchestration to build multiple times, and variant entries in the package json file, unless I'm missing something else.

Hence my comment, I do agree with the opinion of focusing on ESM for now, and adding CJS later if data/feedback requires us to do so.

Reply:
@sebastienlevert  We had this discussion here: 
https://github.com/microsoft/kiota/issues/272#issuecomment-885480031 

Also refer to the following which highlight the struggle with ESM only modules and suggest supporting both: 
https://redfin.engineering/node-modules-at-war-why-commonjs-and-es-modules-cant-get-along-9617135eeca1
https://www.the-guild.dev/blog/support-nodejs-esm

If we are focusing on Node support then we should be focusing on CJS and not ESM.

A good here is node-fetch v3 which migrated to ESM and the move got some heavy feedback https://github.com/node-fetch/node-fetch/issues/1263

Reply:
The `node-fetch` discussion is pretty interesting. How challenging is it to release both? Our focus should be NodeJS for now, but we should be able to "basic" operations within the web, still. If we can simply bundle for both, I'm ok with delivering more value for both audiences (which I understand are very strong and the NodeJS community did not make the full switch to ESM, even if it's supported.

Reply:
The Graph core will continue releasing both as we do today. For the service library I would consider having separate CJS or ESM libraries as an option.